client/src/EntryForm.tsx

```diff
@@ -0,0 +1,144 @@
+import { useEffect, useRef, useState } from 'react';
+import {
+  addEntry,
+  Entry,
+  readEntry,
+  removeEntry,
+  updateEntry,
+} from './data.ts';
+import { useNavigate, useParams } from 'react-router-dom';
+import { createPortal } from 'react-dom';
+
+export function EntryForm() {
+  const [titleText, setTitleText] = useState('');
+  const [urlText, setUrlText] = useState('');
+  const [notesText, setNotesText] = useState('');
+  const navigate = useNavigate();
+  const { entryId } = useParams();
+  const modalRef = useRef<HTMLDialogElement>(null);
+
+  useEffect(() => {
+    async function read() {
+      try {
+        const response = await readEntry(Number(entryId));
+        if (!response) {
+          return;
```

Comment:
You should let the user know that the entryId is invalid. (Maybe they were manually typing numbers into the address bar.) The easiest way to do that is to throw here, instead of return.